### DISCUSSION TITLE
Show and allow the use of actions not in trunk

### DISCUSSION THREAD

**Discussion by Mlim4M** [community/community#11843] [Actions]
When creating a branch, I may require to create new workflows, adding a new workflow to `.github/workflows`. I want to be able to see this workflow when I go to my Actions tab, and I want to be able to execute these new workflows.

It looks like this issue was raised https://github.community/t/workflow-files-only-picked-up-from-master/16129/50 but it seems that either there was a misunderstanding about the original issue, as a timeframe of 2 - 3 weeks lead time was provided to resolve the issue and it is unclear what the resolution of that was.

**Comment by ethomson** [comment]
Can you give me more details about exactly what you're doing - what sort of triggers you're adding to this workflow - and what you expect to happen?  [Our documentation](https://docs.github.com/en/actions/using-workflows/triggering-a-workflow) should indicate the necessary things for a trigger to occur:

> The following steps occur to trigger a workflow run:
> 
> 1. An event occurs on your repository. The event has an associated commit SHA and Git ref.
> 2. GitHub searches the .github/workflows directory in your repository for workflow files that are present in the associated commit SHA or Git ref of the event.
> 3. A workflow run is triggered for any workflows that have on: values that match the triggering event. Some events also require the workflow file to be present on the default branch of the repository in order to run.
> 
>     Each workflow run will use the version of the workflow that is present in the associated commit SHA or Git ref of the event. When a workflow runs, GitHub sets the GITHUB_SHA (commit SHA) and GITHUB_REF (Git ref) environment variables in the runner environment. For more information, see "[Using environment variables](https://docs.github.com/en/actions/automating-your-workflow-with-github-actions/using-environment-variables)."

So I'd like to understand better what type of trigger you're trying to configure, etc.

**Reply by Mlim4M** [reply]
I apologize for not being clear with which triggers I am referring to. 

This is in reference specifically to the `workflow_dispatch` trigger, where I have found that if I have created a new workflow in a feature branch which has a `workflow_dispatch` trigger, I need to push up a dummy workflow into my main trunk to be able to access the `Run workflow` dropdown and select my branch.

The only other alternative to avoid pushing into master is rather then using 'workflow_dispatch' I can use 'push' with my specific branch and hard code the values I need, which feels like a hack.

A simple UI solution I see would be to allow the user to select a specific branch in the Actions screen which then sets the context for what workflows are available, and defaults the filters to that specific branch.

**Reply by nihonjinrxs** [reply]
I'm also dealing with this and it's a very confusing and difficult setup. Developing a workflow to build and test things with a requirement that the workflow already be present on the default branch, which is usually (and in our case) the production deployment branch and quite restricted with regard to pushing/merging commits makes it impossible to ensure that your workflow is testing what you want it to test before merging it. You can work around this by adding a dummy workflow file, but then it becomes unclear what is running at all without diving into the output of each workflow run log. A simple selector "Use workflows from this branch: [ Select a branch... ]" could solve this problem easily, and would allow me to be sure I'm executing the actual workflow I'm developing.

**Reply by philomory** [reply]
This lack of this feature is pretty frustrating, it makes it very hard to test your workflows properly, which you generally want to do before merging them. @nihonjinrxs's suggestion of a branch selector (where you select the branch _before_ you select the workflow) would be very helpful.

**Reply by ethomson** [reply]
Thanks for the additional details @Mlim4M - and the +1s @nihonjinrxs and @philomory. I agree that this is a difficulty, and something that we'll take a look at improving in the future.

**Reply by tyrann0us** [reply]
Hi @ethomson, sorry if this has been discussed elsewhere already, but are there any updates on this?

Given the no-longer-accessible old GitHub community thread with multiple participants (last version [here](https://web.archive.org/web/20220721222645/https://github.community/t/workflow-files-only-picked-up-from-master/16129) on archive.org) and relatively high-rated Stack Overflow questions/answers (examples [here](https://stackoverflow.com/a/67847408) and [here](https://stackoverflow.com/a/65389878)), this is undoubtedly a feature wanted by many GitHub users.

Thanks!

**Reply by tyrann0us** [reply]
Bummer, @ethomson no longer works at GitHub. 🙁 Tagging other GitHub members I've come across.
CC @queenofcorgis, @ossanna16

**Reply by a-matthew** [reply]
> I'm also dealing with this and it's a very confusing and difficult setup. Developing a workflow to build and test things with a requirement that the workflow already be present on the default branch, which is usually (and in our case) the production deployment branch and quite restricted with regard to pushing/merging commits makes it impossible to ensure that your workflow is testing what you want it to test before merging it. You can work around this by adding a dummy workflow file, but then it becomes unclear what is running at all without diving into the output of each workflow run log. A simple selector "Use workflows from this branch: [ Select a branch... ]" could solve this problem easily, and would allow me to be sure I'm executing the actual workflow I'm developing.

Totally agree. It's mind boggling. How can environments be properly tested if an average, Agile Team is expected to use policies prohibiting from pushing into the main/prod branch (under the acceptance criterias, code review etc.) It's quite a normal and non-extravagant assumption to have the dispatchable workflows be available from feature branches directly.

The testing simply can't be done if the requirement enforces prior presence in the main/default branch, as it defeats the purpose of having GitFlow in the first place. The other providers do allow this to work in the expected way.

"A simple selector "Use workflows from this branch: [ Select a branch... ]" could solve this problem easily"
It is exactly how it should work and how it works in different pipelines.

**Comment by Akash1134** [comment]
Hi @tyrann0us, I have asked the engineering team to get more details on this request. 
**Quick Tip** (you might know this already): To see what’s planned next for Actions, visit our [public roadmap](https://github.com/github/roadmap/projects/1?card_filter_query=label%3Aactions)

**Reply by tyrann0us** [reply]
Hi @Akash1134, thanks for getting back to me. The link you posted leads to a closed project. Did you maybe mean this: https://github.com/orgs/github/projects/4247/views/1?filterQuery=label%3Aactions

**Comment by edaniels** [comment]
This is super confusing. We just hit this and only had a clue of the issue being "Some events also require the workflow file to be present on the default branch of the repository in order to run.". At a minimum, please just tell us which events need to be in the workflow branch. GitHub Actions testing is already hard enough as it is!

**Comment by stefan79** [comment]
Is there any update on this?

**Comment by krisdthompson** [comment]
How can we develop in Trunk? Where are my Workflows in my Development branch? An update would be nice.
